File: src/http.ts
import {
  FluxChatApiError,
  FluxChatConfigError,
  FluxChatNetworkError,
} from './errors.js';
import type { FluxChatClientOptions } from './types.js';

// Targets API v2 by default — that's where per-request `context` lives.
const DEFAULT_BASE_URL = 'https://dev-api.fluxchat-corp.com/api/v2';
const DEFAULT_TIMEOUT_MS = 30_000;

interface RequestOptions {
  method: 'GET' | 'POST' | 'PATCH' | 'DELETE';
  path: string;
  body?: unknown;
}

/**
 * Thin fetch wrapper that handles auth, timeouts, error mapping and unwrapping
 * of the FluxChat `{ success, data, timestamp }` response envelope.
 */
export class HttpClient {
  private readonly baseUrl: string;
  private readonly apiKey?: string;
  private readonly token?: string;
  private readonly timeoutMs: number;
  private readonly fetchImpl: typeof fetch;
  private readonly extraHeaders: Record<string, string>;

  constructor(options: FluxChatClientOptions) {
    this.baseUrl = (
      options.baseUrl ??
      (typeof process !== 'undefined' ? process.env?.FLUXCHAT_BASE_URL : undefined) ??
      DEFAULT_BASE_URL
    ).replace(/\/+$/, '');

    this.apiKey = options.apiKey;
    this.token = options.token;
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.extraHeaders = options.headers ?? {};

    const resolvedFetch = options.fetch ?? globalThis.fetch;
    if (!resolvedFetch) {
      throw new FluxChatConfigError(
        'No fetch implementation found. Use Node >= 18 or pass `fetch` in the client options.',
      );
    }
    this.fetchImpl = resolvedFetch;

    if (!this.apiKey && !this.token) {
      throw new FluxChatConfigError(
        'Missing credentials: provide `apiKey` or `token` in the client options.',
      );
    }
  }

  private headers(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      ...this.extraHeaders,
    };
    if (this.apiKey) headers['X-API-Key'] = this.apiKey;
    else if (this.token) headers['Authorization'] = `Bearer ${this.token}`;
    return headers;
  }

  async request<T>({ method, path, body }: RequestOptions): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);

    let response: Response;
    try {
      response = await this.fetchImpl(url, {
        method,
        headers: this.headers(),
        body: body === undefined ? undefined : JSON.stringify(body),
        signal: controller.signal,
      });
    } catch (err) {
      if ((err as Error)?.name === 'AbortError') {
        throw new FluxChatNetworkError(
          `Request to ${path} timed out after ${this.timeoutMs}ms`,
        );
      }
      throw new FluxChatNetworkError(
        `Network error calling ${path}: ${(err as Error).message}`,
      );
    } finally {
      clearTimeout(timer);
    }

    const raw = await response.text();
    const parsed = raw ? safeJsonParse(raw) : undefined;

    if (!response.ok) {
      const message =
        (isRecord(parsed) && typeof parsed.message === 'string' && parsed.message) ||
        `FluxChat API error ${response.status} on ${path}`;
      throw new FluxChatApiError(message, response.status, path, parsed ?? raw);
    }

    // Unwrap the standard envelope: { success, data, timestamp }.
    if (isRecord(parsed) && 'data' in parsed && 'success' in parsed) {
      return parsed.data as T;
    }
    return parsed as T;
  }
}

function safeJsonParse(text: string): unknown {
  try {
    return JSON.parse(text);
  } catch {
    return text;
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}

File: sdk/python/fluxchat/_http.py
"""Internal HTTP helper using httpx."""

from __future__ import annotations
from typing import Any

import httpx

from .exceptions import FluxChatApiError, FluxChatNetworkError


class HttpHelper:
    """Helper HTTP interne — wraps httpx et gère les erreurs."""

    def __init__(self, api_key: str, base_url: str, timeout: float = 30.0,
                 client: httpx.Client | None = None) -> None:
        self._base_url = base_url.rstrip("/")
        self._default_headers = {
            "X-API-Key": api_key,
            "Content-Type": "application/json",
            "Accept": "application/json",
        }
        self._timeout = timeout
        self._client = client or httpx.Client(timeout=timeout)

    def get(self, path: str, extra_headers: dict | None = None) -> Any:
        return self._request("GET", path, extra_headers=extra_headers)

    def post(self, path: str, body: dict | None = None) -> Any:
        return self._request("POST", path, body=body)

    def put(self, path: str, body: dict | None = None) -> Any:
        return self._request("PUT", path, body=body)

    def patch(self, path: str, body: dict | None = None) -> Any:
        return self._request("PATCH", path, body=body)

    def delete(self, path: str) -> None:
        self._request("DELETE", path)

    def _request(self, method: str, path: str,
                 body: dict | None = None,
                 extra_headers: dict | None = None) -> Any:
        headers = {**self._default_headers, **(extra_headers or {})}
        url = self._base_url + path
        try:
            response = self._client.request(
                method, url, headers=headers,
                json=body if body else None,
            )
        except httpx.RequestError as exc:
            raise FluxChatNetworkError(str(exc), cause=exc) from exc

        if not (200 <= response.status_code < 300):
            msg = response.text
            try:
                err_data = response.json()
                if "message" in err_data:
                    msg = err_data["message"]
            except Exception:
                pass
            raise FluxChatApiError(response.status_code, msg)

        if not response.content or response.content == b"null":
            return None

        try:
            data = response.json()
            if isinstance(data, dict) and "data" in data and "success" in data:
                return data["data"]
            return data
        except Exception as exc:
            raise FluxChatNetworkError("Failed to decode JSON response", cause=exc) from exc
